### ISSUE TITLE
[Bash] two bugs in test operator highlighting

### ISSUE THREAD

**Issue by zufuliu** [ScintillaOrg/lexilla#203]
for following code first `-e` is incorrectly highlighted as `SCE_SH_WORD`, second `-e` is not highlighted as `SCE_SH_WORD`.
```bash
[[ $1 == -e* ]] && echo e

if [[ -d /usr/bin &&
	-e /usr/bin/bash ]]; then
	echo find bash
fi
```

Adding extra test for `keywordEnds` fixed first bug, force backtrack for bracket test fixed second bug. 

[bash-test-0905.patch](https://github.com/ScintillaOrg/lexilla/files/12522077/bash-test-0905.patch)
```diff
diff --git a/lexers/LexBash.cxx b/lexers/LexBash.cxx
index d853ce68..b8609d5b 100644
--- a/lexers/LexBash.cxx
+++ b/lexers/LexBash.cxx
@@ -662,7 +662,7 @@ void SCI_METHOD LexerBash::Lex(Sci_PositionU startPos, Sci_Position length, int
 		// handle line continuation, updates per-line stored state
 		if (sc.atLineStart) {
 			CmdState state = CmdState::Body;	// force backtrack while retaining cmdState
-			if (!StyleForceBacktrack(MaskCommand(sc.state))) {
+			if (!StyleForceBacktrack(MaskCommand(sc.state)) && (cmdState != CmdState::Test || testExprType == TestExprType::Test)) {
 				if (!QuoteStack.lineContinuation) {	// retain last line's state
 					cmdState = CmdState::Start;
 				}
@@ -742,7 +742,7 @@ void SCI_METHOD LexerBash::Lex(Sci_PositionU startPos, Sci_Position length, int
 					}
 					// disambiguate option items and file test operators
 					else if (s[0] == '-') {
-						if (cmdState != CmdState::Test || !IsTestOperator(s, setSingleCharOp))
+						if (cmdState != CmdState::Test || !keywordEnds || !IsTestOperator(s, setSingleCharOp))
 							sc.ChangeState(identifierStyle);
 					}
 					// disambiguate keywords and identifiers
```

**Comment by zufuliu** [comment]
I'm planing to remove `testExprType` (by moving `DoubleBracket` and `SingleBracket` into `CmdState`) to simplify the code.

**Comment by zufuliu** [comment]
[bash-test-expr-0908.patch](https://github.com/ScintillaOrg/lexilla/files/12553837/bash-test-expr-0908.patch)
Patch to remove `testExprType`. command delimiters condition changed to `cmdState < CmdState::DoubleBracket`.

**Comment by nyamatongwe** [comment]
So [bash-test-expr-0908.patch](https://github.com/ScintillaOrg/lexilla/files/12553837/bash-test-expr-0908.patch) is just removing `testExprType` preparatory to fixing the stated problem?

**Comment by zufuliu** [comment]
Yes, just removed `testExprType` and updated command delimiter condition.

[bash-test-0908-full.patch](https://github.com/ScintillaOrg/lexilla/files/12554286/bash-test-0908-full.patch)
full patch (with only following additional changes) fix this and others at #197
```c++
// arithmetic expression and double bracket test can span multiline without line continuation
if (!StyleForceBacktrack(sc.state) && !AnyOf(cmdState, CmdState::DoubleBracket, CmdState::Arithmetic)) {


if (!AnyOf(cmdState, CmdState::Test, CmdState::SingleBracket, CmdState::DoubleBracket)
	|| !keywordEnds || !IsTestOperator(s)) {
```

```bash
[[ $1 == -e* ]] && echo e

if [[ -d /usr/bin &&
	-e /usr/bin/bash ]]; then
	echo find bash
fi

echo $((2
- 1))
echo $[2
- 1]
for ((i = 0; i < 2;
	i++)); do
	echo postfix inc $i
done
for ((i = 1; i < 4;
	i<<=1)); do
	echo shift $i
done
```

**Comment by zufuliu** [comment]
[bash-test-full-0908-2.patch](https://github.com/ScintillaOrg/lexilla/files/12558406/bash-test-full-0908-2.patch)
Minor update full patch (moved ` !AnyOf` after `!QuoteStack.lineContinuation`):
```diff
@@ -663,7 +658,9 @@ void SCI_METHOD LexerBash::Lex(Sci_PositionU startPos, Sci_Position length, int
 		if (sc.atLineStart) {
 			CmdState state = CmdState::Body;	// force backtrack while retaining cmdState
 			if (!StyleForceBacktrack(MaskCommand(sc.state))) {
-				if (!QuoteStack.lineContinuation) {	// retain last line's state
+				// retain last line's state
+				// arithmetic expression and double bracket test can span multiline without line continuation
+				if (!QuoteStack.lineContinuation && !AnyOf(cmdState, CmdState::DoubleBracket, CmdState::Arithmetic)) {
 					cmdState = CmdState::Start;
 				}
 				if (QuoteStack.Empty()) {	// force backtrack when nesting
```
